chunky/src/java/se/llbit/chunky/renderer/InternalRenderManager.java

```diff
@@ -0,0 +1,502 @@
+      sceneProvider.withSceneProtected(scene -> {
+        synchronized (bufferedScene) {
+          bufferedScene.copyTransients(scene);
+          updateRenderState(scene);
+        }
+      });
+
+      synchronized (bufferedScene) {
+        bufferedScene.renderTime = System.currentTimeMillis() - renderStart;
+
+        if (snapshotControl.saveSnapshot(bufferedScene, bufferedScene.spp))
+          bufferedScene.postProcessFrame(renderTask);
+        else
+          this.finalizeFrame(false);
+
+        frameCompletionListener.accept(bufferedScene, bufferedScene.spp);
+        updateRenderProgress();
+
+        if (bufferedScene.spp > bufferedScene.getTargetSpp()) {
+          renderCompletionListener.accept(bufferedScene.renderTime, samplesPerSecond());
+          return true;
+        }
+      }
+
+      return mode == RenderMode.PAUSED || sceneProvider.pollSceneStateChange();
+    };
+  }
+
+  /**
+   * This controls most of the render manager logic.
+   */
+  @Override
+  public void run() {
+    try {
+      while (!isInterrupted()) {
+        // Wait for something to happen
+        ResetReason reason = sceneProvider.awaitSceneStateChange();
+
+        // Copy the new scene state to the bufferedScene
+        final boolean[] sceneReset = {false};
+        synchronized (bufferedScene) {
+          sceneProvider.withSceneProtected(scene -> {
+            if (reason.overwriteState()) {
+              bufferedScene.copyState(scene);
+              sceneReset[0] = true;
+            }
+            if (reason == ResetReason.MATERIALS_CHANGED || reason == ResetReason.SCENE_LOADED) {
+              scene.importMaterials();
+              sceneReset[0] = true;
+            }
+
+            bufferedScene.copyTransients(scene);
+            finalizeAllFrames = scene.shouldFinalizeBuffer();
+            updateRenderState(scene);
+
+            if (reason == ResetReason.SCENE_LOADED) {
+              // Make sure frame is finalized
+              bufferedScene.postProcessFrame(renderTask);
+
+              // Swap buffers so the render canvas will see the current frame.
+              bufferedScene.swapBuffers();
+
+              // Notify the scene listeners.
+              sendSceneStatus(bufferedScene.sceneStatus());
+              canvas.repaint();
+            }
+          });
+        }
+
+        // Select the correct renderer
+        Renderer render = mode == RenderMode.PREVIEW ? getPreviewRenderer() : getRenderer();
+
+        if (sceneReset[0]) {
+          render.sceneReset(this, reason);
+        }
+
+        renderStart = System.currentTimeMillis();
+        if (mode == RenderMode.PREVIEW) {
+          // Preview with no CPU limit
+          pool.setCpuLoad(100);
+          render.setPostRender(previewCallback);
+          render.render(this);
+          pool.setCpuLoad(cpuLoad);
+        } else {
+          // Bail early if render is already done
+          if (bufferedScene.spp > bufferedScene.getTargetSpp()) {
```

Comment:
```suggestion
          if (bufferedScene.spp >= bufferedScene.getTargetSpp()) {
```

Or does that break anything (post-processing or something)?

Reply:
Nice catch, mistranslated the original code.